server/src/main/java/org/elasticsearch/transport/OutboundMessage.java

```diff
@@ -41,20 +41,32 @@
     BytesReference serialize(BytesStreamOutput bytesStream) throws IOException {
         storedContext.restore();
         bytesStream.setVersion(version);
-        bytesStream.skip(TcpHeader.HEADER_SIZE);
+        bytesStream.skip(TcpHeader.headerSize(version));
 
         // The compressible bytes stream will not close the underlying bytes stream
         BytesReference reference;
+        int variableHeaderLength;
+        long preHeaderPosition = bytesStream.position();
         try (CompressibleBytesOutputStream stream = new CompressibleBytesOutputStream(bytesStream, TransportStatus.isCompress(status))) {
             stream.setVersion(version);
-            threadContext.writeTo(stream);
+            writeVariableHeader(stream);
```

Comment:
Does it really get us much to store the header size here when we're compressing anyway?
I'm assuming we're doing this to lower the cost of dealing with the header in isolation so we can do smart things with the message at low cost with it.
But even with this change we still have to allocate all the buffers and such for decompression. Maybe we should just pull the header and action out of the compressed section while we're at it? (I believe we talked about it before ... we don't really compress much anyway for the action name and if we do compress a lot for the rest of the header we should probably just change what we serialise there instead of compressing it to make up for redundant stuff?)

Reply:
> But even with this change we still have to allocate all the buffers and such for decompression.

Since we are doing syncFlush we only have to allocate enough buffers for the header. So yes, the idea is to read the header and circuit break / back-off if breaking not available / etc. It kind of will work in conjunction with https://github.com/elastic/elasticsearch/pull/48263 for this functionality.

However, syncFlush does degrade the quality of the compression for the action name / headers. So it is worth questioning if these need to be compressed anyway. We can team-discuss some of these specifics.